pkg/redpanda/client.go

```diff
@@ -0,0 +1,210 @@
+	}
+
+	return kgo.SASL(m)
+}
+
+// DialContextFunc is a function that acts as a dialer for the underlying Kafka client.
+type DialContextFunc func(ctx context.Context, network, host string) (net.Conn, error)
+
+func wrapTLSDialer(dialer DialContextFunc, config *tls.Config) DialContextFunc {
+	return func(ctx context.Context, network, host string) (net.Conn, error) {
+		conn, err := dialer(ctx, network, host)
+		if err != nil {
+			return nil, err
+		}
+		return tls.Client(conn, config), nil
+	}
+}
+
+// KafkaClient creates a client to talk to a Redpanda cluster based on its helm
+// configuration over its internal listeners.
+func KafkaClient(config kube.Config, release helmette.Release, partial redpanda.PartialValues, dialer DialContextFunc, opts ...kgo.Opt) (*kgo.Client, error) {
+	dot, err := redpanda.Dot(release, partial)
+	if err != nil {
+		return nil, err
+	}
+	dot.KubeConfig = config
+
+	values := helmette.Unwrap[redpanda.Values](dot.Values)
+
+	name := redpanda.Fullname(dot)
+	namespace := dot.Release.Namespace
+	serviceName := redpanda.ServiceName(dot)
+	saslUsers := redpanda.SecretSASLUsers(dot)
+	clientCertName := fmt.Sprintf("%s-client", name)
+	kafkaRootCertName := fmt.Sprintf("%s-%s-root-certificate", name, values.Listeners.Kafka.TLS.Cert)
+	serverName := fmt.Sprintf("%s.%s.svc", serviceName, namespace)
+
+	brokers := []string{}
+	for i := int32(0); i < values.Statefulset.Replicas; i++ {
+		brokers = append(brokers, fmt.Sprintf("%s-%d.%s.%s.svc:%d", name, i, serviceName, namespace, values.Listeners.Kafka.Port))
+	}
+
+	opts = append(opts, kgo.SeedBrokers(brokers...))
+
+	serverTLSError := func(err error) error {
+		return fmt.Errorf("error fetching server root CA %s/%s: %w", namespace, kafkaRootCertName, err)
+	}
+	clientTLSError := func(err error) error {
+		return fmt.Errorf("error fetching client certificate default/%s: %w", clientCertName, err)
+	}
+	saslError := func(err error) error {
+		return fmt.Errorf("error fetching SASL authentication for %s/%s: %w", saslUsers.Namespace, saslUsers.Name, err)
+	}
+
+	if redpanda.TLSEnabled(dot) {
+		tlsConfig := &tls.Config{MinVersion: tls.VersionTLS12, ServerName: serverName}
+
+		serverCert, found := helmette.Lookup[corev1.Secret](dot, namespace, kafkaRootCertName)
+		if !found {
+			return nil, serverTLSError(ErrServerCertificateNotFound)
+		}
+
+		serverPublicKey, found := serverCert.Data[corev1.TLSCertKey]
+		if !found {
+			return nil, serverTLSError(ErrServerCertificatePublicKeyNotFound)
+		}
+
+		block, _ := pem.Decode(serverPublicKey)
+		serverParsedCertificate, err := x509.ParseCertificate(block.Bytes)
+		if err != nil {
+			return nil, serverTLSError(fmt.Errorf("unable to parse public key %w", err))
+		}
+		pool := x509.NewCertPool()
+		pool.AddCert(serverParsedCertificate)
+
+		tlsConfig.RootCAs = pool
+
+		if redpanda.ClientAuthRequired(dot) {
+			clientCert, found := helmette.Lookup[corev1.Secret](dot, "default", clientCertName)
```

Comment:
```suggestion
			clientCert, found := helmette.Lookup[corev1.Secret](dot, namespace, clientCertName)
```

Reply:
So, I'm not so sure about this namely because of what I pointed out in Slack (when you were on vacation, you may not have seen it). Found this bit in our chart code which seems to be wrong to me?

https://github.com/redpanda-data/helm-charts/blob/dd21905ce82532acf38680054639710985b3f4e1/charts/redpanda/certs.go#L114-L117

Notice the lack of a `Namespace` there.

Reply:
Ah. I'm 90% certain that's a bug. We should be setting Namespace on all objects. I _think_ that `helm install` implicit sets it if not specified but I could be incorrect there. Do you have a link to the slack thread on hand?

Reply:
Sure, and I hadn't thought about the `helm install` bringing with it implicit namespaces (though if we're going to go about doing directly client-based installation, we should probably either pin the client to a particular namespace, or just explicitly set it on all objects).

Here's the [thread](https://redpandadata.slack.com/archives/C04K2N6S2A3/p1723222386293679) (though no one responded yet since y'all were OOO).

Reply:
So, looks like `helm install` does merge in namespaces implicitly, I'll open up another PR making this explicit here, though I'd definitely be interested in taking a look at the templating tests too to see how we can enforce explicit namespace setting in all of our namespaced resources.